src/builder.rs

```diff
@@ -448,172 +607,254 @@ impl Builder {
-            // specific spent note or output note doesn't reveal anything on its own about
-            // the meaning of that note in the transaction context.
-            spends.shuffle(&mut rng);
-            recipients.shuffle(&mut rng);
-
-            assert_eq!(spends.len(), recipients.len());
-            pre_actions.extend(
-                spends
-                    .into_iter()
-                    .zip(recipients.into_iter())
-                    .map(|(spend, recipient)| ActionInfo::new(spend, recipient, &mut rng)),
-            );
-        }
-
-        // Move some things out of self that we will need.
-        let flags = self.flags;
-        let anchor = self.anchor;
-
-        // Determine the value balance for this bundle, ensuring it is valid.
-        let native_value_balance: V = pre_actions
-            .iter()
-            .filter(|action| action.spend.note.asset().is_native().into())
-            .fold(Some(ValueSum::zero()), |acc, action| {
-                acc? + action.value_sum()
-            })
-            .ok_or(OverflowError)?
-            .into()?;
-
-        // Compute the transaction binding signing key.
-        let bsk = pre_actions
-            .iter()
-            .map(|a| &a.rcv)
-            .sum::<ValueCommitTrapdoor>()
-            .into_bsk();
-
-        // Create the actions.
-        let (actions, circuits): (Vec<_>, Vec<_>) =
-            pre_actions.into_iter().map(|a| a.build(&mut rng)).unzip();
-
-        let bundle = Bundle::from_parts(
-            NonEmpty::from_vec(actions).unwrap(),
-            flags,
-            native_value_balance,
-            self.burn
-                .into_iter()
-                .map(|(asset, value)| Ok((asset, value.into()?)))
-                .collect::<Result<_, BuildError>>()?,
-            anchor,
-            InProgress {
-                proof: Unproven { circuits },
-                sigs: Unauthorized { bsk },
-            },
-        );
-
-        assert_eq!(
-            redpallas::VerificationKey::from(&bundle.authorization().sigs.bsk),
-            bundle.binding_validating_key()
-        );
-        Ok(bundle)
+        rng: impl RngCore,
+    ) -> Result<Option<(UnauthorizedBundle<V>, BundleMetadata)>, BuildError> {
+        bundle(
+            rng,
+            self.anchor,
+            self.bundle_type,
+            self.spends,
+            self.outputs,
+            self.burn,
+        )
     }
 }
 
 /// Partition a list of spends and recipients by note types.
 /// Method creates single dummy ZEC note if spends and recipients are both empty.
+#[allow(clippy::type_complexity)]
 fn partition_by_asset(
     spends: &[SpendInfo],
-    recipients: &[RecipientInfo],
+    outputs: &[OutputInfo],
     rng: &mut impl RngCore,
-) -> HashMap<AssetBase, (Vec<SpendInfo>, Vec<RecipientInfo>)> {
+) -> HashMap<
+    AssetBase,
+    (
+        Vec<(Option<usize>, SpendInfo)>,
+        Vec<(Option<usize>, OutputInfo)>,
+    ),
+> {
```

Comment:
let's add a named type to explain the logic.
let's switch location between the content and the index SpendInfo <-> MetadataIdx
```rust
/// The index of the attached spend or output in the bundle.
/// None indicates a dummy note.
/// The index is used to track the position of the note in the bundle.
type MetadataIdx = Option<usize>;

/// Partition a list of spends and recipients by note types.
/// Method creates single dummy ZEC note if spends and recipients are both empty.
#[allow(clippy::type_complexity)]
fn partition_by_asset(
    spends: &[SpendInfo],
    outputs: &[OutputInfo],
    rng: &mut impl RngCore,
) -> HashMap<
    AssetBase,
    (
        Vec<(SpendInfo, MetadataIdx)>,
        Vec<(OutputInfo, MetadataIdx)>,
    ),
> {
```

Reply:
Done.